core/src/main/scala/kafka/log/Log.scala

```diff
@@ -2254,6 +2256,20 @@ class Log(@volatile var dir: File,
   private[log] def replaceSegments(newSegments: Seq[LogSegment], oldSegments: Seq[LogSegment], isRecoveredSwapFile: Boolean = false): Unit = {
     lock synchronized {
       val sortedNewSegments = newSegments.sortBy(_.baseOffset)
+
+      val (emptyCleanedSegments, nonEmptyCleanedSegments) = sortedNewSegments.partition(_.size == 0)
+
+      if (emptyCleanedSegments.nonEmpty) {
+        val baseOffset = segments.firstKey()
+        if (emptyCleanedSegments.head.baseOffset == baseOffset) {
+          val newLogStartOffset = Math.max(logStartOffset, segments.higherKey(baseOffset))
+          if (highWatermark < newLogStartOffset)
+            maybeIncrementHighWatermark(LogOffsetMetadata(newLogStartOffset))
```

Comment:
We probably don't need this. Currently, the log cleaner never cleans beyond the HWM. If a bug causes the logStartOffset to be larger than HWM, we can just rely on the assertion in maybeIncrementLogStartOffset.

Reply:
Several test cases did not bump the HWM when appending records. The patch will fail such cases. That's why I also update HWM here. Is that okay we could refine those cases then?

Reply:
Hmm, in those test cases, should the HWM be higher to start with? Or is there a legitimate case that the HWM needs to be advanced when log segments are swapped? If it's the former, we will want to patch the test cases.

Reply:
Some tests will update HWM to LEO when they append new records, and some won't. Prior this patch, we don't care about the HWM thing since log start offset does not advance after cleaning. However, with this patch applied, things changed. I think it'd better update HWM to LEO as well for those failed tests. Does it make sense?

Reply:
@huxihx : Thanks for the reply. In general, HWM is supposed to be changed only at the ReplicaManager level since it can only advance if all in-sync replicas have received the records. At the Log layer, we only know the local replica and shouldn't be advancing the HWM directly. That's why I was suggesting to fix the tests instead of changing HWM here so that the tests could pass.